node/subsystem-util/src/database.rs

```diff
@@ -0,0 +1,335 @@
+	use std::{collections::BTreeSet, io::Result};
+
+	/// Adapter implementing subsystem database
+	/// for `KeyValueDB`.
+	#[derive(Clone)]
+	pub struct DbAdapter<D> {
+		db: D,
+		allowed_iter: BTreeSet<u32>,
+	}
+
+	impl<D: KeyValueDB> DbAdapter<D> {
+		/// Instantiate new subsystem database, with
+		/// definition of collections allowing iteration.
+		pub fn new(db: D, allowed_iter: &[u32]) -> Self {
+			DbAdapter { db, allowed_iter: allowed_iter.iter().cloned().collect() }
+		}
+
+		fn filter_iter(&self, col: u32) {
+			debug_assert!(
+				self.is_indexed_column(col),
+				"Invalid configuration of database, column {} is not ordered.",
+				col
+			);
+		}
+
+		fn filter_tx(&self, transaction: &DBTransaction) {
+			debug_assert!({
+				for op in &transaction.ops {
+					if let DBOp::DeletePrefix { col, .. } = op {
+						self.filter_iter(*col);
+					}
+				}
+				true
+			})
+		}
+	}
+
+	impl<D: KeyValueDB> Database for DbAdapter<D> {
+		fn is_indexed_column(&self, col: u32) -> bool {
+			self.allowed_iter.contains(&col)
+		}
+	}
+
+	impl<D: KeyValueDB> KeyValueDB for DbAdapter<D> {
+		fn transaction(&self) -> DBTransaction {
+			self.db.transaction()
+		}
+
+		fn get(&self, col: u32, key: &[u8]) -> Result<Option<DBValue>> {
+			self.db.get(col, key)
+		}
+
+		fn get_by_prefix(&self, col: u32, prefix: &[u8]) -> Option<Box<[u8]>> {
+			self.filter_iter(col);
+			self.db.get_by_prefix(col, prefix)
+		}
+
+		fn write(&self, transaction: DBTransaction) -> Result<()> {
+			self.filter_tx(&transaction);
+			self.db.write(transaction)
+		}
+
+		fn iter<'a>(&'a self, col: u32) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
+			self.filter_iter(col);
+			self.db.iter(col)
+		}
+
+		fn iter_with_prefix<'a>(
+			&'a self,
+			col: u32,
+			prefix: &'a [u8],
+		) -> Box<dyn Iterator<Item = (Box<[u8]>, Box<[u8]>)> + 'a> {
+			self.filter_iter(col);
+			self.db.iter_with_prefix(col, prefix)
+		}
+
+		fn restore(&self, _new_db: &str) -> Result<()> {
+			unimplemented!("restore is unsupported")
+		}
+
+		fn io_stats(&self, kind: IoStatsKind) -> IoStats {
+			self.db.io_stats(kind)
+		}
+
+		fn has_key(&self, col: u32, key: &[u8]) -> Result<bool> {
+			self.db.has_key(col, key)
+		}
+
+		fn has_prefix(&self, col: u32, prefix: &[u8]) -> bool {
+			self.filter_iter(col);
+			self.db.has_prefix(col, prefix)
+		}
+	}
+
+	impl<D: KeyValueDB> MallocSizeOf for DbAdapter<D> {
+		fn size_of(&self, ops: &mut MallocSizeOfOps) -> usize {
+			// ignore filter set
+			self.db.size_of(ops)
+		}
+	}
+}
+
+/// Utilities for using parity-db database.
+pub mod paritydb_impl {
+	use super::{DBTransaction, DBValue, Database, KeyValueDB};
+	use kvdb::{DBOp, IoStats, IoStatsKind};
+	use parity_db::Db;
+	use parking_lot::Mutex;
+	use std::{collections::BTreeSet, io::Result, sync::Arc};
+
+	fn handle_err<T>(result: parity_db::Result<T>) -> T {
+		match result {
+			Ok(r) => r,
+			Err(e) => {
+				panic!("Critical database error: {:?}", e);
```

Comment:
Ouch. I'm not sure this is the best way to handle IO errors. Can we propagate them instead?
We can create a separate trait maybe instead of inheriting KeyValueDB and a different type Iter = impl Iterator<Item=Result<Box<[u8]>, Box<[u8]>, Error>>` or something along the lines akin to https://github.com/paritytech/parity-common/issues/437.

Reply:
Totally agree :), I wanted to keep this pr small and non-obtrusive, but at the same time already added a new db trait.
Question: Should we proceed with trait changes in this PR or later (later being sometime never)?

Reply:
Can be done as a follow-up. But please create an issue so it won't get lost.